File: solutions/javascript/high-scores/1/high-scores.js
/**
 * Class representing a collection of high scores.
 */
export class HighScores {
  /**
   * Create a new instance of HighScores.
   * @param {number[]} scores - The list of scores.
   */
  constructor(scores) {
    /**
     * @property {number[]} scores - The list of scores.
     */
    this.scores = scores;
  }

  /**
   * Get the latest score.
   * @return {number} The latest score.
   */
  get latest() {
    return this.scores[this.scores.length - 1];
  }

  /**
   * Get the personal best score.
   * @return {number} The highest score.
   */
  get personalBest() {
    return Math.max(...this.scores);
  }

  /**
   * Get the top three personal scores.
   * @return {number[]} An array of the top three scores, sorted in descending order.
   */
  get personalTopThree() {
    return [...this.scores]
      .sort((a, b) => b - a)
      .slice(0, 3);
  }
}

File: solutions/lua/high-scores/1/high-scores.lua
local function get_top(max, scores)
  local sorted = {}
  table.move(scores, 1, #scores, 1, sorted)
  table.sort(sorted, function(a, b) return a > b end)
  while #sorted > max do table.remove(sorted) end
  return sorted
end

return function(scores)
  local top = get_top(3, scores)

  return {
    scores = function(self) return scores end,
    latest = function(self) return scores[#scores] end,
    personal_best = function(self) return top[1] end,
    personal_top_three = function(self) return top end
  }
end